File: 0572-subtree-of-another-tree/0572-subtree-of-another-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isSubtree(TreeNode root, TreeNode subRoot) {
        
        if(root == null)
            return false;
                
        if(subRoot.val == root.val){
            if(isTreeSame(root.left, subRoot.left) && isTreeSame(root.right, subRoot.right))            
                return true;
        }
                
        return isSubtree(root.left, subRoot) || isSubtree(root.right, subRoot);
    }
    
    public boolean isTreeSame(TreeNode root, TreeNode subRoot){
        
        if(root == null && subRoot == null)
            return true;
        
        if((root != null && subRoot == null) || 
          (root == null && subRoot != null)){
            return false;
        }
        
        if(root.val != subRoot.val){
            return false;
        }
        
        if(isTreeSame(root.left, subRoot.left) &&
          isTreeSame(root.right, subRoot.right))
            return true;
        
        return false;
    }
}

File: 0572-subtree-of-another-tree/0572-subtree-of-another-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:

        if root is None and subRoot:
            return False
        
        if root and subRoot is None:
            return False
        
        if root is None and subRoot is None:
            return True

        if self.isIdentical(root, subRoot):
            return True

        return self.isSubtree(root.left, subRoot) or self.isSubtree(root.right, subRoot)

    def isIdentical(self, root, subRoot):

        if root is None and subRoot is None:
            return True

        if root is None or subRoot is None:
            return False

        if root.val != subRoot.val:
            return False

        return self.isIdentical(root.left, subRoot.left) and self.isIdentical(root.right, subRoot.right)